identity_stardust/src/state_metadata_document.rs

```diff
@@ -0,0 +1,326 @@
+use serde::Serialize;
+
+use crate::did_or_placeholder::DIDOrPlaceholder;
+use crate::did_or_placeholder::PLACEHOLDER_DID;
+use crate::StardustDocument;
+
+/// The DID document as it is contained in the state metadata of an alias output.
+#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
+pub struct StateMetadataDocument(CoreDocument<DIDOrPlaceholder>);
+
+impl StateMetadataDocument {
+  /// Transforms the document into a [`StardustDocument`] by replacing all placeholders with `original_did`.
+  pub fn into_stardust_document(mut self, original_did: &CoreDID) -> StardustDocument {
+    let mut builder: DocumentBuilder<CoreDID, Object, Object, Object> =
+      CoreDocument::builder(self.0.properties().clone()).id(original_did.clone());
+
+    let mut controllers = None;
+    std::mem::swap(&mut controllers, self.0.controller_mut());
+
+    let original_did_ctor = || original_did.clone();
+
+    if let Some(controllers) = controllers {
+      for controller in controllers.into_vec().into_iter() {
+        builder = builder.controller(controller.unwrap_or_else(original_did_ctor));
+      }
+    }
+
+    for method in self
+      .0
+      .verification_method_mut()
+      .as_vec_mut()
+      .drain(..)
+      .map(|method| method.map(|did| did.unwrap_or_else(original_did_ctor)))
+    {
+      builder = builder.verification_method(method);
+    }
+
+    let method_ref_transform =
+      |method_ref: MethodRef<DIDOrPlaceholder, _>| method_ref.map(|did| did.unwrap_or_else(original_did_ctor));
+
+    for method_ref in self
+      .0
+      .assertion_method_mut()
+      .as_vec_mut()
+      .drain(..)
+      .map(method_ref_transform)
+    {
+      builder = builder.assertion_method(method_ref);
+    }
+    for method_ref in self
+      .0
+      .capability_delegation_mut()
+      .as_vec_mut()
+      .drain(..)
+      .map(method_ref_transform)
+    {
+      builder = builder.capability_delegation(method_ref);
+    }
+    for method_ref in self
+      .0
+      .capability_invocation_mut()
+      .as_vec_mut()
+      .drain(..)
+      .map(method_ref_transform)
+    {
+      builder = builder.capability_invocation(method_ref);
+    }
+    for method_ref in self
+      .0
+      .key_agreement_mut()
+      .as_vec_mut()
+      .drain(..)
+      .map(method_ref_transform)
+    {
+      builder = builder.key_agreement(method_ref);
+    }
+    for method_ref in self
+      .0
+      .authentication_mut()
+      .as_vec_mut()
+      .drain(..)
+      .map(method_ref_transform)
+    {
+      builder = builder.authentication(method_ref);
+    }
+
+    for service in self.0.service_mut().as_vec_mut().drain(..) {
+      builder = builder.service(service.map(|did| did.unwrap_or_else(original_did_ctor)));
+    }
+
+    let mut core_doc: CoreDocument = builder.build().expect("all required document fields should be set");
+
+    std::mem::swap(core_doc.also_known_as_mut(), self.0.also_known_as_mut());
+
+    StardustDocument(core_doc)
```

Comment:
All of this is unnecessary. This is why the `map` function exists.

 ```suggestion
     let mut core_doc: CoreDocument<DIDOrPlaceholder> = self.0;
    let core_document: CoreDocument = core_doc.map(
      // Replace placeholder identifiers.
      |did| match did {
        DIDOrPlaceholder::Placeholder => original_did.clone(),
        DIDOrPlaceholder::Core(core) => core,
      },
      // Do not modify properties.
      |o| o,
    );
    StardustDocument(core_document)
```